Make the wxART_* constants const wxStrings with static storage duration

StefanBruens opened the pull request:
The pre-3.1.4 definitions of the constants were just string literals,
while the type were wxString typedefs. To avoid implicit conversion
these were converted to actual wxStrings.

While the interface now matches the implementation, this has several
drawbacks:
- every use of the "constant" now is a string construction at runtime
- the constant now is an rvalue, i.e. it is impossible to take its
  address.

The latter breaks its use from wxPython.

See also https://github.com/wxWidgets/wxWidgets/pull/1312

StefanBruens committed:
Make the wxART_* constants const wxStrings with static storage duration

The pre-3.1.4 definitions of the constants were just string literals,
while the type were wxString typedefs. To avoid implicit conversion
these were converted to actual wxStrings.

While the interface now matched the implementation, this has several
drawbacks:
- every use of the "constant" now is a string construction at runtime
- the constant now is an rvalue, i.e. it is impossible to take its
  address.

The latter breaks its use from wxPython.

The IDs are moved to a separate file which can be included multiple
times, once from the header to have the declarations in place, and once
to instantiate the wxStrings. Using a common file avoids the declaration
and definition going out of sync.

ardovm commented:
Apparently, some builds are failing because some of the the `wxArtClient` and `wxArtId` instances are not being included in the object file.
Please try moving the line:
```c++
#define WX_DEFINE_WX_ART_IDS 1
```
_before_ the inclusion of `wx/wxprec.h`

StefanBruens commented:
> Apparently, some builds are failing because some of the the `wxArtClient` and `wxArtId` instances are not being included in the object file.
> Please try moving the line:
> 
> ```c++
> #define WX_DEFINE_WX_ART_IDS 1
> ```
> 
> _before_ the inclusion of `wx/wxprec.h`

This would definitely be wrong.

https://gcc.gnu.org/onlinedocs/gcc/Precompiled-Headers.html
> Any macros defined before the precompiled header is included must either be defined in the same way as when the precompiled header was generated, or must not affect the precompiled header, which usually means that they don’t appear in the precompiled header at all.

The precompiled header can either contain only the declarations (then you end up with missing symbols), or also with the definitions, which would violate the ODR.

One possible solution is to move the IDs to a separate header file, _without_ any include guards. This file can then be included multiple times. Having identical declarations multiple times is not a problems, and the definition would only be in artprov.cpp.

vadz commented:
I didn't think at all about taking address of these constants, but if we really have to support this, we must indeed have actual `wxString` variables for the art IDs and the current change works for this, thanks!

I'd like to rename the new file, however: we don't really use `_` in the file names (with a few legacy exceptions), perhaps we could call it just `artids.h`?

Alternatively, we could get rid of the macros in the header and just have `extern wxString wxART_XXX;` in it and matching `wxDEFINE_ART_ID(wxART_XXX);` in the source file. This does mean that we'd have to duplicate all IDs, but they change very rarely and I wonder if it wouldn't be better to just do it instead of complicating things with X-macros. What do you think?

StefanBruens commented:
> I'd like to rename the new file, however: we don't really use `_` in the file names (with a few legacy exceptions), perhaps we could call it just `artids.h`?

Done.

vadz commented:
Thanks for the update!

I initially thought that it was a fluke, but the segmentation fault in GTK 2 unit tests seems persistent, as it happened again for the new version. Have you tried reproducing it locally and could you see it if you have?

Also, the more I think about it, the more it seems like it would be best to just declare/define variables without using any macros (at least not in the header, where it would also help with various tools parsing them), what do the others think?

StefanBruens commented:
> Also, the more I think about it, the more it seems like it would be best to just declare/define variables without using any macros (at least not in the header, where it would also help with various tools parsing them), what do the others think?

Only `interfaces/artprov.h` is processed, e.g. by doxygen.

StefanBruens commented:
> I initially thought that it was a fluke, but the segmentation fault in GTK 2 unit tests seems persistent, as it happened again for the new version. Have you tried reproducing it locally and could you see it if you have?

I have tested the GTK3 variant, with the wxPhoenix test suite and by running KiCad, no problems.

ardovm commented:
> I initially thought that it was a fluke, but the segmentation fault in GTK 2 unit tests seems persistent, as it happened again for the new version. Have you tried reproducing it locally and could you see it if you have?

On my system (Ubuntu 18.04.4 LTS x86_64) I tried to reproduce the problem.
I configure'd with command:
```../configure wxGTK_VERSION=2 wxUSE_XVFB=1```
and launched the test with and without xvfb:
```sh
cd tests
$ export LD_LIBRARY_PATH=../lib
./test_gui
xvfb-run -a -s '-screen 0 1600x1200x24' ./test_gui
```
Both runs did not crash, but rather showed some failures.
Do you think I am missing anything in order to reproduce the problem?

> Also, the more I think about it, the more it seems like it would be best to just declare/define variables without using any macros (at least not in the header, where it would also help with various tools parsing them), what do the others think?

IMHO it would be also better to have declarations in the header and definitions in the CPP file, simply entered one by one, without macros.
@StefanBruens 's idea of a macro to change the effect of the header (declaring or defining) is very smart, but the combination with precompiled headers makes the problem too complex.

ardovm commented:
> I initially thought that it was a fluke, but the segmentation fault in GTK 2 unit tests seems persistent, as it happened again for the new version. Have you tried reproducing it locally and could you see it if you have?

I downloaded a Trusty image from osboxes.org and tried to reproduce the problem there. It's systematic!
I think there is a race condition among static variable initializers, in which the `wxId2Gtk` array in file `src/gtk/artgtk.cpp` is initialized before the `wxArtID`'s it requires.

```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff507bdeb in std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::basic_string(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&) ()

(gdb) bt
#0  0x00007ffff507bdeb in std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >::basic_string(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007ffff5fa489e in wxString::wxString (
    this=0x7ffff672c960 <(anonymous namespace)::wxId2Gtk>, stringSrc=...)
    at ../include/wx/string.h:1140
#2  0x00007ffff6046f6d in __static_initialization_and_destruction_0 (
    __initialize_p=1, __priority=65535) at ../src/gtk/artgtk.cpp:125
#3  0x00007ffff6047960 in _GLOBAL__sub_I_artgtk.cpp(void) ()
    at ../src/gtk/artgtk.cpp:404
#4  0x00007ffff7dea1da in call_init (l=<optimized out>, argc=argc@entry=1, 
    argv=argv@entry=0x7fffffffdf88, env=env@entry=0x7fffffffdf98)
    at dl-init.c:78
#5  0x00007ffff7dea2c3 in call_init (env=<optimized out>, 
    argv=<optimized out>, argc=<optimized out>, l=<optimized out>)
    at dl-init.c:36
#6  _dl_init (main_map=0x7ffff7ffe1c8, argc=1, argv=0x7fffffffdf88, 
    env=0x7fffffffdf98) at dl-init.c:126
#7  0x00007ffff7ddb29a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
[...]
```

Inspecting frame 1:
```
(gdb) frame 1
#1  0x00007ffff5fa489e in wxString::wxString (
    this=0x7ffff672c960 <(anonymous namespace)::wxId2Gtk>, stringSrc=...)
    at ../include/wx/string.h:1140
1140      wxString(const wxString& stringSrc) : m_impl(stringSrc.m_impl) { }
```
where `stringSrc` seems to be actually `wxART_ERROR`:
```
(gdb) print stringSrc
$6 = (const wxString &) @0xa3ef40: {static npos = 18446744073709551615, 
  m_impl = <error reading variable: Cannot access memory at address 0xffffffffffffffe8>, m_convertedToChar = {m_str = 0x0, m_len = 0}}

(gdb) print &wxART_ERROR
$8 = (const wxArtID *) 0xa3ef40 <wxART_ERROR@@WXU_3.1>
```

This probably never happened before, because plain `const char *` did not need any initialization.

The quickest solution, IMHO, is to refactor the `wxId2Gtk` array, using _pointers_ to `wxArtId`. It does not seem to be used anywhere outside of that file.

vadz commented:
Oh, of course, thanks for finding this problem. The standard solution for this kind of things is to use accessor functions instead of variables. E.g.
```cpp
// artprov.h
const wxString& wxArtIdFoo();
#define wxART_FOO wxArtIdFoo()

// artprov.cpp
const wxString& wxArtIdFoo()
{
    static const s_foo("FOO");
    return s_foo;
}
```

StefanBruens commented:
> Oh, of course, thanks for finding this problem. The standard solution for this kind of things is to use accessor functions instead of variables. E.g.
> 
> ```c++
> // artprov.h
> const wxString& wxArtIdFoo();
> #define wxART_FOO wxArtIdFoo()
> 
> // artprov.cpp
> const wxString& wxArtIdFoo()
> {
>     static const s_foo("FOO");
>     return s_foo;
> }
> ```

There a two possibilities to write this (https://isocpp.org/wiki/faq/ctors#static-init-order):

```
    static const wxString* sp_foo("FOO");
    return *sp_foo;
```
*or*
```
    static const wxString s_foo("FOO");
    return s_foo;
```
The second is only safe if
a) none of the ART ids are accessed after teardown starts (not only initialization order is undefined, but also deinitialization order)
or b) the destructor does not invalidate the contents of the wxString

b) is likely not the case, as at least for a wchar_t-unicode build the string data points to runtime-allocated data*

[*] side note - for unicode builts, "FOO" should be changed to "FOO"L

vadz commented:
> There a two possibilities to write this (https://isocpp.org/wiki/faq/ctors#static-init-order):
> 
> ```
>     static const wxString* sp_foo("FOO");
>     return *sp_foo;
> ```
> 
> _or_
> 
> ```
>     static const wxString s_foo("FOO");
>     return s_foo;
> ```
> 
> The second is only safe if
> a) none of the ART ids are accessed after teardown starts (not only initialization order is undefined, but also deinitialization order)
> or b) the destructor does not invalidate the contents of the wxString

This should be perfectly safe here because art IDs can't be used before the library is initialized or after it is shut down.

> [*] side note - for unicode builts, "FOO" should be changed to "FOO"L

It should actually be `wxS("FOO")` if we want to avoid run-time conversions in UTF-8 builds too.

StefanBruens commented:
> This should be perfectly safe here because art IDs can't be used before the library is initialized or after it is shut down.

This is apparently not as clear as you think - usage obviously happens during initialization (otherwise we would not see the crash), and if there is a single destructor which uses an art ID, you are back to sqare one.

StefanBruens commented:
```
// artprov.h
const wxString& wxArtIdFoo();
#define wxART_FOO wxArtIdFoo()
```
The problem with this is, it again changes the type of wxART_FOO - it no longer is a wxString, but a function call, i.e. it no longer matches the interface, e.g. `const wxArtID wxART_ADD_BOOKMARK;`.

vadz commented:
> > This should be perfectly safe here because art IDs can't be used before the library is initialized or after it is shut down.
> 
> This is apparently not as clear as you think - usage obviously happens during initialization (otherwise we would not see the crash), and if there is a single destructor which uses an art ID, you are back to sqare one.

You're right, I've completely misdiagnosed the problem. It's not due to art IDs being initialized too late, as I assumed for some reason, but due to them being initialized too early, before the library itself is initialized. This is not allowed and we shouldn't do this.

Of course, the solution is still the same: by using wrapper functions we avoid creating the objects until they're really needed, which should only happen once the initialization is complete.

> The problem with this is, it again changes the type of wxART_FOO - it no longer is a wxString, but a function call, i.e. it no longer matches the interface, e.g. const wxArtID wxART_ADD_BOOKMARK;.

Why is this a problem as long as you can still write `&wxART_FOO`? Interface didn't match the actual type before neither...

StefanBruens commented:
> The problem with this is, it again changes the type of wxART_FOO - it no longer is a wxString, but a function call, i.e. it no longer matches the interface, e.g. `const wxArtID wxART_ADD_BOOKMARK;`.

The initialization order problem can also be solved by using an accessor function in artgk.cpp for the mapping table.

StefanBruens commented:
> > The problem with this is, it again changes the type of wxART_FOO - it no longer is a wxString, but a function call, i.e. it no longer matches the interface, e.g. const wxArtID wxART_ADD_BOOKMARK;.
> 
> Why is this a problem as long as you can still write `&wxART_FOO`? Interface didn't match the actual type before neither...

Because e.g. wxPython cheated and rewrote the type:

https://github.com/wxWidgets/Phoenix/blob/8b670176d2cc9d84ee656c05d15d0cd5a026902b/etg/artprov.py#L55-L62

vadz commented:
So it seems like wxPython will be changed anyhow, because these symbols definitely can't be just `const char*` any longer (otherwise passing them to functions taking `wxString` wouldn't work any longer when implicit conversions are disabled). Could the code at your link be changed to use `const wxString&` type for them?

StefanBruens commented:
Just removing the type reassignment from the wxPython code-generator-generator works fine, it keeps  the 'const wxPython' type from the interface then.

You can not use the wxString& returned from an accessor function here, again it would be trying to  take the address of an rvalue (which does not work, even if the reference is pointing to a static object).

StefanBruens commented:
> > This is apparently not as clear as you think - usage obviously happens during initialization (otherwise we would not see the crash), and if there is a single destructor which uses an art ID, you are back to sqare one.
> 
> You're right, I've completely misdiagnosed the problem. It's not due to art IDs being initialized too late, as I assumed for some reason, but due to them being initialized too early, before the library itself is initialized. This is not allowed and we shouldn't do this.
> 
> Of course, the solution is still the same: by using wrapper functions we avoid creating the objects until they're really needed, which should only happen once the initialization is complete.

The art IDs _are_ initialized to too late. We have two static objects, both are created by the runtime linker when the library is loaded. The linker calls the constructor of the wxArtID (for each one, actually), and it calls the constructor of the mapping table in artgtk. The mapping table constructor calls the copy constructor of each wxString/wxArtId, which may be after *or before* the wxArtId constructor. The sequence of the two constructors is undefined, and we have *no influence on the order*.

StefanBruens commented:
Another approach is to make `wxString::wxString(const underlying_storage_type*)` a `constexpr`, which then guarantees by C++11 rules e.g. `const wxString wxART_FOO(wxS("FOO"))` is const-initialized, i.e. before any other static initialization.

StefanBruens committed:
Use local static variable for ArtID mapping table

The instantiation of the global static mapping table may happen before
the instantiation of the wxArtIds, i.e. the copy-constructor would
access not yet valid wxArtId/wxString instances.

Delay the instantiation until the first call to wxArtIDToStock call.
This is still not 100% safe, as wxArtIDToStock may be called by a
constructor of a static instance, but hopefully nobody does that.

ardovm commented:
Just giving my two cents.

@StefanBruens 
> > The problem with this is, it again changes the type of wxART_FOO - it no longer is a wxString, but a function call, i.e. it no longer matches the interface, e.g. `const wxArtID wxART_ADD_BOOKMARK;`.
> 
> The initialization order problem can also be solved by using an accessor function in artgk.cpp for the mapping table.

There may be an even simpler solution: change the `wxId2Gtk` array to be a static variable _inside_ function `wxArtIDToStock()` because it only seems to be accessed in there.
In this way, the array will be initialized only when the function `wxArtIDToStock()` is first called, and I believe this will only happen after static initialization has been carried out.

@vadz 
> So it seems like wxPython will be changed anyhow, because these symbols definitely can't be just const char* any longer (otherwise passing them to functions taking wxString wouldn't work any longer when implicit conversions are disabled). Could the code at your link be changed to use const wxString& type for them?

In theory, wxPython developers will be able to choose whether the implicit conversion are enabled or disabled when compiling their own code, so this should not affect them. Right?
I cannot think of any use cases of other C++ code being compiled together with wxPython, therefore the above choice would not affect anyone else.

vadz commented:
Sorry, I still don't understand what is the problem with returning `const wxString&` (no, it isn't rvalue), could you please explain what exactly doesn't work if we did this?

Also, I agree that we should use a wrapper function for `wxId2Gtk` initialization anyhow, there is absolutely no reason to do it during startup when we can safely postpone it for later (and maybe avoid doing it at all). Doing this will fix the crash, but we could still have a similar problem in the user code, which is why using wrapper functions for `wxART_FOO` still looks like a good idea to me.

vadz commented on src/gtk/artgtk.cpp:
```diff
@@ -122,27 +133,23 @@ const wxString wxId2Gtk[] = {
     // allow passing GTK+ stock IDs to wxArtProvider -- if a recognized wx
     // ID wasn't found, pass it to GTK+ in the hope it is a GTK+ or theme
     // icon name:
     wxString ret(id);
```
This is unused now and should be removed (or the change to the last line of this function should be undone and `break` in the loop be restored).

vadz commented on src/gtk/artgtk.cpp:
```diff
@@ -122,27 +133,23 @@ const wxString wxId2Gtk[] = {
         }
     }
-    return ret;
+    return wxString(id);
```
```suggestion
    return id;
```

vadz commented on src/gtk/artgtk.cpp:
```diff
@@ -58,7 +58,18 @@ namespace
 #endif
 
-const wxString wxId2Gtk[] = {
+namespace {
```
This is unnecessary, we're already inside an anonymous namespace here.
```suggestion
```

vadz commented on src/gtk/artgtk.cpp:
```diff
@@ -58,7 +58,18 @@ namespace
 
-const wxString wxId2Gtk[] = {
+namespace {
+struct wxArtStockMapping
```
I'd declare this inside the function below to make it even more local.

vadz commented on src/gtk/artgtk.cpp:
```diff
@@ -58,7 +58,18 @@ namespace
+
+wxString wxArtIDToStock(const wxArtID& id)
+{
+  static const wxArtStockMapping wxId2GtkMap[] =
```
It's a bit confusing to call something which is an array a "map"... BTW, we could actually use `wxStringToStringHashMap` here.

vadz commented on src/gtk/artgtk.cpp:
```diff
@@ -58,7 +58,18 @@ namespace
     #define ART(wxId, stockId, unused) wxId, stockId,
```
This should bechanged to  `wxArtStockMapping(wxId, stockId),` now if only to avoid `-Wmissing-braces` warnings (but also for clarity).

vadz reviewed (commented):
Thanks for the update, as I wrote, I think this should be enough to fix the test failure, but I'd still like to change `wxART_FOO` to be functions rather than objects.

vadz commented:
The latest version broke the build in C++98 mode...

vadz commented:
Note to self: update `build/files` when merging.

vadz committed:
Remove trailing blank line from the new wx/artids.h file

No real changes.

vadz committed:
Add ctor to wxArtStockMapping to fix C++98 build

Can't use aggregate initialization without C++11.

vadz committed:
Add wx/artids.h to build/files and rerun upmake and bakefile

Finish adding the new file, which was previously added manually to a
couple of files, but not everywhere.

vadz committed:
Remove useless "Modified by:" header line

This is not needed nor used since a long time any longer.

vadz commented:
I've tried to fix the build and a couple of other minor issues, will merge if the CI builds finally pass.
